src/features/auth/RequireAuth.tsx

```diff
@@ -0,0 +1,51 @@
+import { ReactElement, useEffect } from 'react';
+
+import { LoadingModal } from 'components';
+import { useWeb3React } from 'hooks/useWeb3React';
+
+import { useAuthStep } from './useAuthStep';
+import { useFinishAuth } from './useFinishAuth';
+import { useWalletAuth } from './useWalletAuth';
+import { WalletAuthModal } from './WalletAuthModal';
+
+export const RequireAuth = (props: { children: ReactElement }) => {
+  const { address, authTokens } = useWalletAuth();
+  const web3Context = useWeb3React();
+  const finishAuth = useFinishAuth();
+  const [authStep, setAuthStep] = useAuthStep();
+
+  useEffect(() => {
+    // reset after logging out or signature error
+    if (authStep !== 'connect' && !web3Context.active) {
+      setAuthStep('connect');
+      return;
+    }
+
+    if (authStep === 'connect' && web3Context.active) {
+      setAuthStep('sign');
+      finishAuth({ web3Context, authTokens })
+        .then(success => {
+          if (!success) {
```

Comment:
nit: assuming success is a bool, then this is fine... but i generally prefer to case on the affirmative, and else on the not affirmative for a bit more readability, and safer handling if there are every multiple cases that could be returned.

```suggestion
          if (success) {
                      setAuthStep('done');
                      }
          else { 
          web3Context.deactivate();
          }
```